### ISSUE TITLE
Phase 2: SFT Saturation Study - How pretraining duration affects fine-tuning

### ISSUE THREAD

**Issue by toddwbucy** [toddwbucy/Atlas-MAG_OmegaRule#45]
## Research Question

Does extended pretraining (beyond Chinchilla optimal) improve SFT efficiency and final quality?

## Hypothesis

Models trained longer on pretraining data may:
- Converge faster during SFT
- Achieve better final SFT quality
- Show different memory module utilization patterns

### Checkpoints to Evaluate
Sample checkpoints from Chinchilla optimal (4,400) through end of training (18,000):
```
4400, 5000, 5500, 6000, 6500, 7000, 7500, 8000, 8500, 9000, ...
```
~28 checkpoints total

### Parallel Execution
- **GPU 0**: SFT checkpoint A
- **GPU 2**: SFT checkpoint B
- Process pairs in parallel → 14 runs total

### Metrics to Capture
- [ ] SFT convergence speed (steps to reach target loss)
- [ ] Final SFT loss
- [ ] Downstream benchmark scores
- [ ] NIAH accuracy on SFT'd models
- [ ] Memory vs attention contribution post-SFT

### SFT Dataset
TBD - needs to be consistent across all checkpoints

## Automation Requirements

Build a generalizable checkpoint→SFT pipeline:

```
inputs:
  - checkpoint_dir: path to pretrained checkpoints
  - sft_dataset: path or HF dataset name
  - gpu_ids: [0, 2]
  - sft_config: hyperparameters

outputs:
  - sft_models/: fine-tuned checkpoints
  - results.jsonl: metrics per checkpoint
  - plots/: convergence curves, saturation analysis
```

## Generalizability

This tool is **not Atlas-specific**. Core components:
1. Checkpoint discovery and scheduling
2. Parallel GPU job dispatch
3. SFT training loop (standard HF/custom)
4. Metrics collection and aggregation
5. Visualization

Could be packaged as a standalone CLI:
```bash
sft-sweep --checkpoints runs/*/checkpoint_*.pt \
          --dataset alpaca \
          --gpus 0,2 \
          --output results/
```

## Success Criteria

- [ ] Automated pipeline runs without intervention
- [ ] Clear curve: pretraining tokens vs SFT performance
- [ ] Identify optimal pretraining duration for this model size
- [ ] Reusable tool for future experiments

## Dependencies

- Pretraining run completion (issue #43 ✅)
- SFT dataset selection
- Benchmark suite selection

---

Phase 2 of Atlas-MAG research. Builds on Phase 1 (architecture validation).
